src/manifest.py

```diff
@@ -486,10 +486,15 @@ def _update_appdata(self):
             elif data.source_path == self._root_manifest_path:
                 selected_data = data
         else:
-            # Guess that the last external source in the root manifest is the one
-            # corresponding to the main application bundle.
-            assert selected_data is not None
-            log.warning("Guessed upstream source: %s", selected_data)
+            if selected_data is not None:
+                # Guess that the last external source in the root manifest is the one
+                # corresponding to the main application bundle.
+                log.warning("Guessed last source as main source: %s", selected_data)
+            else:
+                log.error(
+                    "The last source is not external. Appdata will not be updated"
```

Comment:
Looking at https://github.com/flathub/com.riverbankcomputing.PyQt.BaseApp/blob/branch/6.4/com.riverbankcomputing.PyQt.BaseApp.yaml I don't think this error message is correct.

The condition is that there is no external source in the root manifest.

```suggestion
                log.warning(
                    "No main source configured; and no external source in %s. Not updating appdata.",
                    self._root_manifest_path,
```

Reply:
Seems fine.

Reply:
Actually saying there is no external source seems wrong? The manifest can have external source but if there is no `is-main-source` and the last one is not external, it will fail?

"No external source" would mean there was no external source anywhere.

Reply:
slightly changed it.

Reply:
> The manifest can have external source but if there is no is-main-source and the last one is not external, it will fail?

The loop runs across all external sources in order of appearance. If a source has the main-source property, `selected_data` is set and the loop terminated early. Otherwise, if the source is in the root manifest, `selected_data` is updated. Finally, if the loop has not terminated early, this block runs.

So if this block is reached, if there is 1 or more external source in the root manifest, `selected_data` will be non-`None`; otherwise it will be `None`.

So I think it is correct to say that `selected_data` will be `None` in this block only if there are *no* external sources in the root manifest.

Reply:
Hm ok I thought the bug happened even with an external source early (with no is-main-source) and a non external one as the last one.

I'll change it back then.